force-app/main/service/DonationHistoryService.cls

```diff
@@ -0,0 +1,176 @@
+        List<DonationHistoryController.DonationHistory> result = verifyCreditCondition(donations);
+        return result;
+    }
+
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description method to return total number of record
+    */
+    public static List<AggregateResult> getTotalRecords(Id contactId){
+        List<AggregateResult> totalRecords = [SELECT COUNT(Id)
+                                FROM OpportunityContactRole 
+                                WHERE Opportunity.IsWon = true 
+                                AND ContactId =:contactId];
+        return totalRecords;
+    }
+
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Verify each condition to list Donation.
+    */
+    private static List<DonationHistoryController.DonationHistory> verifyCreditCondition(List<OpportunityContactRole> donations){
+        List<DonationHistoryController.DonationHistory> donationHistory = new List<DonationHistoryController.DonationHistory>();
+        List<DonationHistoryController.DonationHistory> hardCredit = verifyHardCreditCondition(donations);
+        List<DonationHistoryController.DonationHistory> softCredit = verifySoftCreditCondition(donations);
+        List<Id> donationsId = filterVerifiedIds(softCredit);
+        List<DonationHistoryController.DonationHistory> partialSoftCredit = verifyPartialSoftCreditCondition(donationsId, donations);
+        List<DonationHistoryController.DonationHistory> partialAndSoftCredit = verifyPartialAndSoftCreditCondition(softCredit, partialSoftCredit);
+        
+        donationHistory.addAll(hardCredit);
+        donationHistory.addAll(partialAndSoftCredit);
+
+        return donationHistory;
+    }
+    
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Verify hard credit condition.
+    */
+    private static List<DonationHistoryController.DonationHistory> verifyHardCreditCondition(List<OpportunityContactRole> donations){
+        List<DonationHistoryController.DonationHistory> hardCreditList = new List<DonationHistoryController.DonationHistory>();
+        for(OpportunityContactRole ocr: donations){
+            if (ocr.isPrimary) {
+                DonationHistoryController.DonationHistory hardCredit = new DonationHistoryController.DonationHistory(ocr, HARD_CREDIT);
+                hardCreditList.add(hardCredit);
+            }
+        }
+        return hardCreditList;
+    }
+    
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Verify soft credit condition.
+    */
+    private static List<DonationHistoryController.DonationHistory> verifySoftCreditCondition(List<OpportunityContactRole> donations){
+        //with the list of donations verify if the Role match the filter rule value on the NPSP_ClosedWon_Opps_SC filter group
+        Filter_Rule__mdt roles = [SELECT Constant__c FROM Filter_Rule__mdt WHERE QualifiedApiName = 'NPSP_OppsSoftCreditRoles'];
```

Comment:
One more extremely important point - in a unit test context, there will be zero records in the `Filter_Rule__mdt` object **and** it's not possible to insert records into that object. 

See the `CMT_FilterRule_TEST.mockFilterCMTValues()` method for a pattern on how to mock data in that object. Once you follow those steps, you'll need to modify this method use the existing `CMT_FilterRule_SEL` class to retrieve the mocked records - maybe with the `getFilterRules()` method and then loop through the collection to find the record with the defined qualified APi name.

Also, I recommend moving 'NPSP_OppsSoftCreditRoles' to a constant in this class.

Reply:
After checking and debugging the content of `Filter_Rule__mdt`, I can confirm it returns data in the test context. 
Given this facts do you still think the suggested changes above are required?
I will put NPSP_OppsSoftCreditRoles in a constant!

Reply:
You're testing that in your personal scratch org where customizable rollups is already enabled and thus that object has data. That is not the case in a new scratch org, which is exactly how our automated builds function. In that context, the object will have zero records and this will fail.

Reply:
One more note on these, instead of doing String.split(';') on the Constant__c field, you can construct an instance of a ` new CMT_FilterRule(Filter_Rule__mdt filterRule)`. Then inside of the loop, you can call `isIncluded(SObject)` on it, and pass in the OCR you want to evaluate against that rule.

Reply:
When testing this implementation with the instance of `CMT_FilterRule`, it throws an error when passing as a parameter the ocr record to `isIncluded(SObject)` method. Since that should be resolved in the `CMT_FilterRule` class, for this time we keep the split(';')